
<h1 align="center">Peer-graded Assignment: Segmenting and Clustering </h1>
<h1 align="center">Neighborhoods in Toronto</h1>

### Notebook contains the final assignment for week 3 of the Applied Data Science Capstone.

#### Greg Waitt
#### 3/24/21

### Required packages for assignment

In [6]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

# Part 1 - Create DataFrame with PostalCode, Borough, Neighborhood

### Importing Data

In [7]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
html_data = requests.get(url).text

In [8]:
soup = BeautifulSoup(html_data, 'html5lib')

In [9]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

In [10]:
# print(table_contents)
df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [11]:
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

Response 200 means that we are able to make the connection to the page

Group df by PostalCode, reset index, and remove "Index" column

In [12]:
df = df.groupby(['PostalCode']).head()
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [13]:
df = df.reset_index(drop=True)
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [14]:
df.shape

(103, 3)

### Part 1 Answer: 103 rows and 3 columns

# Part 2  Latitude and Lognitude Coordinates for Neighborhoods

### Required packages for assignment

In [16]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

check the shape of the two dataframes we need to join

In [22]:
df.shape

(103, 3)

In [23]:
data.shape

(103, 3)

Let's make sure the Postal Code columns are the same data type

In [24]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [25]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [28]:
corr_data = df.join(data.set_index('Postal Code'), on='PostalCode', how='inner')
corr_data.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [29]:
corr_data.shape

(103, 5)

### Part 2 Answer: 103 rows and 5 columns